### PR TITLE
Add Normal Field Error as a plot quantity to `plot_2d` and `plot_3d` and wrapper for Bplasma from virtual casing

### LINKED ISSUE #910: plotting function for B-normal
Basically a way to visualize the errors returned by `QuadraticFlux.compute` from #726 

Want 2D and 3D options, similar to `plot_2d` and `plot_3d`

### LINKED ISSUE #965: Make wrapper function for computing the virtual casing biot savart integral field on the plasma surface
Should have a higher level wrapper to do the following code (calulcating the keys, making the interpolator, then passing those into the virtual casing biot savart function) (taken from `compute_Bnormal`)

```python
   # do virtual casing to find plasma contribution to Bnormal
            data_keys = [
                "K_vc",
                "B",
                "R",
                "phi",
                "Z",
                "e^rho",
                "n_rho",
                "|e_theta x e_zeta|",
            ]
            vc_eval_data = eq.compute(data_keys, grid=eval_grid)
            vc_data = eq.compute(data_keys, grid=vc_source_grid)

            k = min(
                vc_source_grid.num_theta, vc_source_grid.num_zeta * vc_source_grid.NFP
            )
            s = k - 1
            k = min(
                vc_source_grid.num_theta, vc_source_grid.num_zeta * vc_source_grid.NFP
            )
            q = k // 2 + int(np.sqrt(k))
            try:
                interpolator = FFTInterpolator(eval_grid, vc_source_grid, s, q)
            except AssertionError as e:
                print(
                    f"Unable to create FFTInterpolator, got error {e},"
                    "falling back to DFT method which is much slower"
                )
                interpolator = DFTInterpolator(eval_grid, vc_source_grid, s, q)
            if hasattr(eq.surface, "Phi_mn"):
                vc_data["K_vc"] += eq.surface.compute("K", grid=vc_source_grid)["K"]
            Bplasma = virtual_casing_biot_savart(
                vc_eval_data,
                vc_data,
                interpolator,
            )
            # need extra factor of B/2 bc we're evaluating on plasma surface
            Bplasma = Bplasma + vc_eval_data["B"] / 2
            Bnorm += np.sum(Bplasma * vc_eval_data["n_rho"], axis=-1)
```

### PR DESCRIPTION
**Pull request by dpanici**
- Add `"B*n"` as a plottable quantity to `plot_2d` and `plot_3d`. If passed in, then you must also pass in a `field` kwarg that is a `MagneticField` object and optionally a `field_grid` kwarg as well for the source grid of the magnetic field.

Resolves #910 
Resolves #965

### CODE REVIEW

File: desc/plotting.py [hunk 1]
```diff
@@ -626,8 +630,28 @@ def plot_2d(
     plot_axes = _get_plot_axes(grid)
     if len(plot_axes) != 2:
         return ValueError(colored("Grid must be 2D", "red"))
-
-    data, label = _compute(eq, name, grid, kwargs.pop("component", None))
+    component = kwargs.pop("component", None)
+    if name != "Bn":
```
Comment (ddudt) [inline desc/plotting.py:634]: Can we change the name to `"B*n"` to better match our existing format for other compute quantities? For example, we already have stuff like `"B*grad(|B|)"`

File: desc/plotting.py [hunk 2]
```diff
@@ -599,6 +599,10 @@ def plot_2d(
         * ``ylabel_fontsize``: float, fontsize of the ylabel
         * ``cmap``: str, matplotlib colormap scheme to use, passed to ax.contourf
         * ``levels``: int or array-like, passed to contourf
+        * ``field``: MagneticField, a magnetic field with which to calculate Bn on
```
Comment (ddudt) [inline desc/plotting.py:602]: Idea: to better match the Objectives API, we could change the first argument `eq` to `things`, which would normally be the equilibrium but in this case would be the tuple `(eq, field)`. Instead of adding a new kwarg. And this would also be more general for adding new plot quantities in the future.

File: desc/plotting.py [hunk 3]
```diff
@@ -626,8 +630,28 @@ def plot_2d(
...
+            ValueError,
+            "If Bn is entered as the variable to plot, a magnetic field"
+            " must be provided.",
+        )
+        field_grid = kwargs.pop("field_grid", None)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            data, _ = field.compute_Bnormal(eq, eval_grid=grid, source_grid=field_grid)
+        data = data.reshape((grid.num_theta, grid.num_rho, grid.num_zeta), order="F")
+        label = r"$|B_n| ~(\mathrm{T})$"
```
Comment (ddudt) [inline desc/plotting.py:654]: similar to the above comment, change this label to `r"$|\mathbf{B}\cdot\hat{n}| ~(\mathrm{T})$"` to better match our existing compute quantities

File: desc/plotting.py [hunk 4]
```diff
@@ -851,18 +879,32 @@ def plot_3d(
     levels = kwargs.pop("levels", None)
     component = kwargs.pop("component", None)
 
+    if name != "Bn":
```
Comment (ddudt) [inline desc/plotting.py:882]: same comments for `plot_3d` as for `plot_2d`

**Review by ddudt — CHANGES_REQUESTED** [review]

File: desc/plotting.py [hunk 5]
```diff
@@ -626,8 +630,28 @@ def plot_2d(
...
+        errorif(
+            field is None,
+            ValueError,
+            "If Bn is entered as the variable to plot, a magnetic field"
+            " must be provided.",
+        )
+        field_grid = kwargs.pop("field_grid", None)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            data, _ = field.compute_Bnormal(eq, eval_grid=grid, source_grid=field_grid)
```
Comment (f0uriest) [inline desc/plotting.py:652]: will this only be correct for vacuum equilibria? What if is a nonozero Bn from the plasma field?

**Review by f0uriest — COMMENTED** [review]
should figure out what to do for finite plasma field.

Also might be worth adding an option to normalize by total B?

File: desc/plotting.py (hunk 5 above)
Comment (dpanici) [reply desc/plotting.py:657]: Right now have it using the same plot grid for eval and source, I figure most times it will be fine. Can change it to be separate grids if needed

File: desc/plotting.py (hunk 2 above)
Comment (dpanici) [reply desc/plotting.py:607]: This is a good idea, but not necessarily needed by this PR. I will make an issue for it

File: desc/plotting.py (hunk 5 above)
Comment (f0uriest) [reply desc/plotting.py:657]: I think same grid is probably fine, the plotting resolution is already usually much higher than what we use during optimization.

File: desc/plotting.py [hunk 6]
```diff
@@ -626,8 +635,62 @@ def plot_2d(
...
+                    "R",
+                    "phi",
+                    "Z",
+                    "e^rho",
+                    "n_rho",
+                    "|e_theta x e_zeta|",
+                ],
+                grid=grid,
+            )
+            k = min(grid.num_theta, grid.num_zeta * grid.NFP)
```
Comment (f0uriest) [inline desc/plotting.py:671]: remind me to make a higher level wrapper to do this sort of stuff for common use cases

File: desc/plotting.py [hunk 7]
```diff
@@ -626,8 +635,62 @@ def plot_2d(
...
+            except AssertionError:
+                interpolator = DFTInterpolator(grid, grid, s, q)
+            if hasattr(eq.surface, "Phi_mn"):
+                vc_data["K_vc"] += eq.surface.compute("K", grid=grid)["K"]
+            Bplasma = virtual_casing_biot_savart(
+                vc_data,
+                vc_data,
+                interpolator,
+            )
+            # need extra factor of B/2 bc we're evaluating on plasma surface
```
Comment (f0uriest) [inline desc/plotting.py:686]: technically we don't need this since the equilibrium B always has zero normal component

**Review by ddudt — CHANGES_REQUESTED** [review]
I don't think this is working as intended for `plot_3d`. This is the plot I get for the new test function `test_3d_plot_Bn`: 

plot_3d_Bn

**Comment by dpanici** [comment]
> I don't think this is working as intended for `plot_3d`. This is the plot I get for the new test function `test_3d_plot_Bn`:
> 
> plot_3d_Bn

THis is due to the default grid having endpoint=True, while the singular integral needs endpoint=False to work properly. I will fix this

**Comment by dpanici** [comment]
put in compute bnormal function

File: desc/plotting.py (hunk 7 above)
Comment (dpanici) [reply desc/plotting.py]: Ah is this only needed for the pressure balance?

Comment (f0uriest) [reply desc/plotting.py]: yeah

File: desc/magnetic_fields/_core.py [hunk 8]
```diff
@@ -252,9 +273,64 @@ def compute_Bnormal(
...
+                    "K_vc",
+                    "B",
+                    "R",
+                    "phi",
+                    "Z",
+                    "e^rho",
+                    "n_rho",
+                    "|e_theta x e_zeta|",
+                ],
+                grid=vc_source_grid,
```
Comment (f0uriest) [inline desc/magnetic_fields/_core.py:304]: think we need an actual default for this, otherwise it will use a quadrature grid

File: desc/plotting.py [hunk 9]
```diff
@@ -626,8 +630,55 @@ def plot_2d(
...
+                    NFP=grid.NFP,
+                    endpoint=False,
+                )
+            else:
+                vc_grid = grid
+            data, _ = field.compute_Bnormal(
+                eq, eval_grid=vc_grid, source_grid=field_grid, vc_source_grid=vc_grid
+            )
+        data = data.reshape((vc_grid.num_theta, vc_grid.num_zeta), order="F")
+        if grid.endpoint:
```
Comment (f0uriest) [inline desc/plotting.py:675]: shouldnt this be `if not grid.endpoint`?

File: desc/plotting.py [hunk 10]
```diff
@@ -626,8 +630,55 @@ def plot_2d(
...
+            "If B*n is entered as the variable to plot, "
+            "the grid nodes must be at rho=1.",
+        )
+
+        field_grid = kwargs.pop("field_grid", None)
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+
+            if grid.endpoint:
+                print("changing grids")
```
Comment (f0uriest) [inline desc/plotting.py:661]: rogue print statement

File: desc/plotting.py [hunk 11]
```diff
@@ -851,18 +906,57 @@ def plot_3d(
...
+                    NFP=grid.NFP,
+                    endpoint=False,
+                )
+            else:
+                vc_grid = grid
+            data, _ = field.compute_Bnormal(
+                eq, eval_grid=vc_grid, source_grid=field_grid, vc_source_grid=vc_grid
+            )
+        data = data.reshape((vc_grid.num_theta, vc_grid.num_zeta), order="F")
+        if grid.endpoint:
```
Comment (f0uriest) [inline desc/plotting.py:949]: see above

File: tests/test_plotting.py [hunk 12]
```diff
@@ -1005,3 +1005,30 @@ def test_plot_omnigenous_field():
     )
     fig, ax = plot_boozer_surface(field, iota=0.6, fieldlines=4)
     return fig
+
+
+@pytest.mark.mpl_image_compare(remove_text=True, tolerance=tol_2d)
+@pytest.mark.unit
+def test_2d_plot_Bn():
+    """Test 2d plotting of Bn on equilibrium surface."""
+    eq = get("precise_QA")
```
Comment (f0uriest) [inline tests/test_plotting.py:1014]: recommend using a lower resolution equilibrium here

File: desc/magnetic_fields/_core.py [hunk 13]
```diff
@@ -238,23 +256,91 @@ def compute_Bnormal(
...
-            surface = surface.surface
+            calc_Bplasma = True
+            eq = surface
+            surface = eq.surface
         if eval_grid is None:
             eval_grid = LinearGrid(
                 rho=jnp.array(1.0), M=2 * surface.M, N=2 * surface.N, NFP=surface.NFP
             )
+        if vc_source_grid is None:
+            source_NFP = surface.NFP if surface.N > 0 else 64
```
Comment (dpanici) [inline desc/magnetic_fields/_core.py:271]: Question on this, when N of the eq is zero, we make the virtual casing source grid NFP be 64. But if the eq N is 0 then so is the eq N_grid right? so how is it, for example in the tutorial axisymmetric free boundary solve, that the toroidal effects are accounted for if the source grid has no toroidal grid nodes?

Comment (f0uriest) [reply desc/magnetic_fields/_core.py:271]: there's always still nodes at zeta=0, and NFP=64 means we repeat this 64 times around the torus. There's no toroidal coupling though so each of those can be done independently.

File: docs/notebooks/tutorials/free_boundary_equilibrium.ipynb
Comment (f0uriest) [inline docs/notebooks/tutorials/free_boundary_equilibrium.ipynb:1]: what was changed here?

Comment (f0uriest) [reply docs/notebooks/tutorials/free_boundary_equilibrium.ipynb:1]: nvm see the new plot

**Review by f0uriest — DISMISSED** [review]

**Review by ddudt — APPROVED** [review]
I made a formatting change to reduce the number of lines of code, and also changed to `endpoint=True` in the 3d plot test since that's the more common use case and covers more of the code

**Review by f0uriest — DISMISSED** [review]

**Review by f0uriest — APPROVED** [review]
